promql/functions.go

```diff
@@ -474,6 +474,43 @@ func funcDeriv(ev *evaluator, args Expressions) Value {
 	return resultVector
 }
 
+// === predict(node ExprMatrix, k ExprScalar) Vector ===
+func funcPredict(ev *evaluator, args Expressions) Value {
+	vector := funcDeriv(ev, args[0:1]).(Vector)
+	duration := clientmodel.SampleValue(clientmodel.SampleValue(ev.evalFloat(args[1])))
+
+	excludedLabels := map[clientmodel.LabelName]struct{}{
+		clientmodel.MetricNameLabel: {},
+	}
+
+	// Calculate predicted delta over the duration.
+	signatureToValue := map[uint64]clientmodel.SampleValue{}
+	for _, el := range vector {
+		signature := clientmodel.SignatureWithoutLabels(el.Metric.Metric, excludedLabels)
+		signatureToValue[signature] = el.Value * duration
+	}
+
+	// add predicted delta to last value.
+	matrixBounds := ev.evalMatrixBounds(args[0])
+	outVec := Vector{}
```

Comment:
If you do `outVec := make(Vector, len(signatureToValue))`, that should preallocate the map optimally for the number of expected elements.